rts/motoko-rts/src/memory/ic.rs

```diff
@@ -75,26 +77,29 @@ impl Memory for IcMemory {
         let new_hp = old_hp + delta;
 
         // Grow memory if needed
-        grow_memory(new_hp);
+        if (old_hp ^ new_hp) >> 16 != 0 {
```

Comment:
Just an idea (to be tested for performance):

Maybe, we could introduce a constant `PAGE_SIZE: usize = 65 * 1024;` and derive the check and the last page boundary from that. 

`if (new_hp - old_hp) / PAGE_SIZE > 0 {`

Reply:
(Most probably, the compiler will optimize the division to the bit shift.)

Reply:
See also `WASM_PAGE_SIZE` in line 4. I'll do this experiment.

Reply:
``` rust
if (old_hp ^ new_hp) >> 16 != 0 {
```
already produces
``` wasm
  (func $alloc_words (type 8) (param i32) (result i32)
    (local i32 i64 i64)
    block  ;; label = @1
      global.get $__memory_base
      i32.const 71584
      i32.add
      i32.load
      local.tee 1
      i64.extend_i32_u
      local.tee 2
      local.get 0
      i32.const 2
      i32.shl
      i64.extend_i32_u
      i64.add
      local.tee 3
      local.get 2
      i64.xor
      i64.const 65536
      i64.lt_u
      br_if 0 (;@1;)
      local.get 3
      call $motoko_rts::memory::ic::grow_memory::hb0efe958cfedcb7b
    end
    global.get $__memory_base
    i32.const 71584
    i32.add
    local.get 3
    i64.store32
    local.get 1
    i32.const -1
    i32.add)
```
When having
``` rust
        if (old_hp ^ new_hp) / WASM_PAGE_SIZE != 0 {
```
I get
``` wasm
  (func $alloc_words (type 8) (param i32) (result i32)
    (local i32 i64 i64)
    block  ;; label = @1
      global.get $__memory_base
      i32.const 71584
      i32.add
      i32.load
      local.tee 1
      i64.extend_i32_u
      local.tee 2
      local.get 0
      i32.const 2
      i32.shl
      i64.extend_i32_u
      i64.add
      local.tee 3
      local.get 2
      i64.xor
      i64.const 65536
      i64.lt_u
      br_if 0 (;@1;)
      local.get 3
      call $motoko_rts::memory::ic::grow_memory::hb0efe958cfedcb7b
    end
    global.get $__memory_base
    i32.const 71584
    i32.add
    local.get 3
    i64.store32
    local.get 1
    i32.const -1
    i32.add)
```

Reply:
Except luc was asking for a clearer test. sans XOR, I think:

`if (new_hp - old_hp) / PAGE_SIZE > 0 {`

Reply:
Except that wouldn't work :-) E.g. `old_hp == 0xA_FFF0` and `new_hp == 0xB_0008`, the `new_hp - old_hp` is then `0x18` and if we divide by 0x1_0000` we get 0.

Reply:
Oh, yes, true, sorry for the confusion.
It would be `if (new_hp / PAGE_SIZE > old_hp / PAGE_SIZE)`
But this is maybe slower?

Reply:
I am pretty sure that is slower :-)